syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+{
+	int res;
+	uint8_t id = 0;
+	unsigned int i;
+
+	for (i = 0; i < nb; ++i) {
+		if ((res = sysgen_mapNameResolve(mapNames, &id)) < 0) {
+			fprintf(stderr, "\nCan't add map %s", mapNames);
+			return res;
+		}
+
+		mapIDs[i] = id;
+		mapNames += strlen(mapNames) + 1; /* name + '\0' */
+	}
+
+	return 0;
+}
+
+
+static size_t sysgen_mapsParse(char *maps, char sep)
+{
+	size_t nb = 0;
+
+	while (*maps != '\0') {
+		if (*maps == sep) {
+			*maps = '\0';
+			++nb;
+		}
+		maps++;
+	}
+
+	return ++nb;
+}
+
+
+static int sysgen_appAdd(const char *name, char *imaps, char *dmaps, const char *appArgv, uint32_t flags)
+{
+	int res;
+	char *argv;
+	sysptr_t ptr;
+	struct phfs_alias_t *alias;
+	syspage_prog_t *prog, *headProg;
+
+	size_t dmapSz, imapSz, len, argvSz;
+	const uint32_t isExec = (flags & flagSyspageExec) != 0;
+
+	alias = sysgen_aliasFind(name);
+	if (alias == NULL)
+		return -1;
+
+	/* First instance in imap is a map for the instructions */
+	imapSz = sysgen_mapsParse(imaps, ';');
+	dmapSz = sysgen_mapsParse(dmaps, ';');
+
+	len = strlen(appArgv);
+	argvSz = isExec + len + 1; /* [X] + argv + '\0' */
+
+	ptr = sysgen_buffAlloc(sizeof(syspage_prog_t));
+	if (ptr == 0)
+		return -1;
+
+	prog = CAST_SYSPTR(syspage_prog_t *, ptr);
+	prog->dmaps = sysgen_buffAlloc(sizeof(uint8_t) * dmapSz);
+	prog->imaps = sysgen_buffAlloc(sizeof(uint8_t) * imapSz);
+	prog->argv = sysgen_buffAlloc(sizeof(uint8_t) * argvSz);
+
+	if ((prog->dmaps == 0) || (prog->imaps == 0) || (prog->argv == 0))
+		return -1;
+
+	prog->imapSz = imapSz;
+	prog->dmapSz = dmapSz;
+	prog->start = alias->addr;
+	prog->end = alias->addr + alias->size;
+
+	argv = CAST_SYSPTR(char *, prog->argv);
+	if (isExec)
+		argv[0] = 'X';
+
+	memcpy(argv + isExec, appArgv, len);
+	argv[argvSz - 1] = '\0';
+
+	if ((res = sysgen_mapsAdd2Prog(CAST_SYSPTR(uint8_t *, prog->imaps), imapSz, imaps)) < 0 ||
+		(res = sysgen_mapsAdd2Prog(CAST_SYSPTR(uint8_t *, prog->dmaps), dmapSz, dmaps)) < 0)
+		return res;
+
+	if (sysgen_common.syspage->progs == 0) {
+		prog->next = ptr;
+		prog->prev = ptr;
+		sysgen_common.syspage->progs = ptr;
+	}
+	else {
+		headProg = CAST_SYSPTR(syspage_prog_t *, sysgen_common.syspage->progs);
+
+		prog->prev = headProg->prev;
+		CAST_SYSPTR(syspage_prog_t *, headProg->prev)->next = ptr;
+		prog->next = sysgen_common.syspage->progs;
+		headProg->prev = ptr;
+	}
+
+	return 0;
+}
+
+
+int sysgen_cmdApp(int argc, char *argv[])
+{
+	size_t pos;
+	int argvID = 0;
+
+	char *imaps, *dmaps;
+	unsigned int flags = 0;
+
+	const char *appArgv;
+	char name[SIZE_CMD_ARG_LINE];
+
+	if (argc < 5 || argc > 6) {
+		fprintf(stderr, "\n%s: Wrong argument count", argv[0]);
+		return -1;
+	}
+
+	/* ARG_0: command name */
+
+	/* ARG_1: alias to device - it will be checked in phfs_open */
+
+	/* ARG_2: optional flags */
+	argvID = 2;
+	if (argv[argvID][0] == '-') {
+		if ((argv[argvID][1] | 0x20) == 'x' && argv[argvID][2] == '\0') {
```

Comment:
What does | 0x20 is supposed to do on a character?

Reply:
Thanks to it, user can define uppercase or lowercase in script.